File: engine/cover/render.mjs
// Headless render of the weekly cover, using the site's own renderer so the
// engine and the lab can never disagree. No dev server needed: the renderer
// source is injected into a blank page, Figtree is loaded from Google Fonts,
// and both variants (with the name for Spotify, plain for the site) are
// written as JPEG under the Spotify 256 KB limit.
//
//   node engine/cover/render.mjs --name "Held Breath" --issue 2 \
//     --palette "#c08a66,#ee2519,#317a97,#882229" --recipe '{"blur":40}' \
//     --out-text site/public/issues/cover-002.jpg --out-plain site/public/issues/cover-002-plain.jpg
//
// PLAYWRIGHT_DIR overrides where the playwright package lives (default: the
// machine's existing install under listen-logger/node_modules).

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

const here = path.dirname(fileURLToPath(import.meta.url));
const args = Object.fromEntries(process.argv.slice(2).reduce((acc, a, i, arr) => {
  if (a.startsWith("--")) acc.push([a.slice(2), arr[i + 1] && !arr[i + 1].startsWith("--") ? arr[i + 1] : "true"]);
  return acc;
}, []));
// Playwright comes from whichever app has it installed (reader/ or site/
// after `npm install`), or from PLAYWRIGHT_DIR. Its own downloaded Chromium
// is used unless PLAYWRIGHT_CHROMIUM points at another binary.
const ROOT = path.resolve(here, "../..");
const PW_CANDIDATES = [process.env.PLAYWRIGHT_DIR, path.join(ROOT, "reader/node_modules/playwright"), path.join(ROOT, "site/node_modules/playwright")].filter(Boolean);
const PW_DIR = PW_CANDIDATES.find((p) => fs.existsSync(path.join(p, "index.mjs")));
if (!PW_DIR) { console.error("playwright not found: run `npm install` in reader/ (then `npx playwright install chromium`), or set PLAYWRIGHT_DIR"); process.exit(1); }
const EXE = process.env.PLAYWRIGHT_CHROMIUM || undefined;
const { chromium } = await import(pathToFileURL(path.join(PW_DIR, "index.mjs")).href);

const size = parseInt(args.size || "640", 10);
const name = args.name || "";
const issue = parseInt(args.issue || "0", 10);
const palette = (args.palette || "").split(",").map((s) => s.trim()).filter(Boolean);
const recipe = args.recipe ? JSON.parse(args.recipe) : {};
const maxBytes = parseInt(args["max-bytes"] || String(256 * 1024), 10);
// The cover painter lives with the frozen first reader (the redesign
// promotion moved site/src; the archive copy is identical).
const src = fs.readFileSync(path.resolve(here, "../../site/reader/cover/render.js"), "utf8") + "\nwindow.__renderCover = renderCover;\n";

const browser = await chromium.launch({ executablePath: fs.existsSync(EXE) ? EXE : undefined });
const page = await browser.newPage({ viewport: { width: size + 40, height: size + 40 }, deviceScaleFactor: 1 });
await page.setContent(`<!doctype html><html><head><meta charset="utf-8">
<link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Figtree:wght@500;600&display=swap">
<style>body{margin:0;background:#fff;font-family:Figtree,sans-serif}</style></head>
<body><span style="font:600 20px Figtree">warm</span><canvas id="c"></canvas></body></html>`, { waitUntil: "load" });
await page.addScriptTag({ type: "module", content: src });
await page.waitForFunction(() => typeof window.__renderCover === "function", null, { timeout: 10000 });
await page.evaluate(async () => {
  try { await Promise.all([document.fonts.load("600 64px Figtree"), document.fonts.load("500 20px Figtree")]); await document.fonts.ready; } catch {}
});

async function render(plain) {
  for (const q of [0.9, 0.86, 0.82, 0.78, 0.72, 0.66]) {
    const dataUrl = await page.evaluate(({ name, issue, palette, recipe, size, plain, q }) => {
      const c = document.getElementById("c");
      window.__renderCover(c, { name, issue, palette, recipe, size, plain });
      return c.toDataURL("image/jpeg", q);
    }, { name, issue, palette, recipe, size, plain, q });
    const buf = Buffer.from(dataUrl.split(",")[1], "base64");
    if (buf.length <= maxBytes || q === 0.66) return { buf, q };
  }
}
const out = {};
if (args["out-text"]) { const { buf, q } = await render(false); fs.writeFileSync(args["out-text"], buf); out.text = { path: args["out-text"], bytes: buf.length, quality: q }; }
if (args["out-plain"]) { const { buf, q } = await render(true); fs.writeFileSync(args["out-plain"], buf); out.plain = { path: args["out-plain"], bytes: buf.length, quality: q }; }
await browser.close();
console.log(JSON.stringify(out));

File: engine/cover/render.py
"""Render the week's two covers and record them on the issue.

  cover-NNN.jpg        the gradient mesh with the issue's name (Spotify)
  cover-NNN-plain.jpg  the same mesh without text (site list view)

Inputs: the issue's name, its extracted palette (engine.cover.palette) and
the recipe saved from the cover lab (site/cover-recipe.json; the renderer's
defaults when absent). Rendering is the site's own canvas code run headless
(render.mjs), so what you tuned in the lab is exactly what ships. Records
companion_playlist.cover_art on the issue (issues/ and the site copy).

Run:  PYTHONPATH=<root> python3 -m engine.cover.render --n 2
"""

import argparse
import json
import shutil
import subprocess
import sys
from pathlib import Path

from ..lib.config import PROJECT_ROOT, load_config

NEUTRAL = ["#1a1a1a", "#6e6e6e", "#c9c9c9", "#f4f4f4"]


def display_title(issue):
    t = issue.get("title") or f"Issue {issue.get('issue', 0):03d}"
    return t.split(": ", 1)[1] if ": " in t else t


def main(argv=None):
    ap = argparse.ArgumentParser(description=__doc__.split("\n")[0])
    ap.add_argument("--n", type=int, required=True)
    ap.add_argument("--size", type=int, default=640)
    args = ap.parse_args(argv)
    cfg = load_config()
    name = f"issue-{args.n:03d}.json"
    src = Path(cfg["ISSUES_DIR"]) / name
    if not src.exists():
        raise SystemExit(f"{src} not found")
    node = shutil.which("node")
    if not node:
        print("cover render: node not found; skipping")
        return 0
    issue = json.loads(src.read_text())
    site_dir = Path(cfg["SITE_ISSUES_DIR"])
    pal_p = site_dir / f"palette-{args.n:03d}.json"
    palette = NEUTRAL
    if pal_p.exists():
        try:
            colors = json.loads(pal_p.read_text()).get("colors") or []
            palette = colors if colors else NEUTRAL
        except Exception:
            pass
    recipe = {}
    rec_p = PROJECT_ROOT / "site" / "cover-recipe.json"
    source = "week"
    if rec_p.exists():
        try:
            doc = json.loads(rec_p.read_text())
            recipe = doc.get("recipe") or {}
            source = doc.get("source") or "week"
            if source == "custom" and doc.get("custom"):
                palette = doc["custom"]
        except Exception:
            pass
    out_text = site_dir / f"cover-{args.n:03d}.jpg"
    out_plain = site_dir / f"cover-{args.n:03d}-plain.jpg"
    cmd = [node, str(Path(__file__).with_name("render.mjs")),
           "--name", display_title(issue), "--issue", str(args.n),
           "--palette", ",".join(palette), "--recipe", json.dumps(recipe),
           "--size", str(args.size), "--out-text", str(out_text), "--out-plain", str(out_plain)]
    res = subprocess.run(cmd, capture_output=True, text=True, timeout=180)
    if res.returncode != 0:
        print("cover render failed:", (res.stderr or res.stdout).strip()[:600])
        return 1
    info = json.loads(res.stdout.strip().splitlines()[-1])
    rel = {"text": f"/issues/{out_text.name}", "plain": f"/issues/{out_plain.name}",
           "palette": palette, "recipe_source": source}
    for p in (src, site_dir / name):
        d = json.loads(p.read_text())
        d.setdefault("companion_playlist", {})["cover_art"] = rel
        p.write_text(json.dumps(d, indent=2, ensure_ascii=False) + "\n")
    print(f"cover {args.n:03d}: {out_text.name} {info['text']['bytes']//1024} KB (q{info['text']['quality']}), "
          f"{out_plain.name} {info['plain']['bytes']//1024} KB; recorded on the issue")
    return 0


if __name__ == "__main__":
    sys.exit(main())
